### PR TITLE
Revert "Revert frozen RuntimeType objects and write barriers for frozen objects"

### PR DESCRIPTION
**Pull request by EgorBo**
This PR reverts https://github.com/dotnet/runtime/pull/76235 with a few manual modifications:
1) JIT-EE guid is re-generated
2) `gen.bat` is invoked (ThunkInput.txt)
3) Crst are re-generated by the tool, changed lock-level to Unordered

The rest of the changes are automatically reverted.

The main issue why the initial PRs (https://github.com/dotnet/runtime/pull/75573 and https://github.com/dotnet/runtime/pull/76135) were reverted has just been resolved via https://github.com/dotnet/runtime/pull/76251

tldr:
GC could collect some associated (with frozen objects) objects as unreachable, e.g. it could collect a SyncBlock, WeakReferences and Dependent handles associated with frozen objects which could (e.g. for a short period of time) be indeed unreachable but return back to life after.

Unblocks https://github.com/dotnet/runtime/pull/76112

### CODE REVIEW

File: src/coreclr/jit/ee_il_dll.cpp [hunk 1]
```diff
@@ -1615,45 +1615,40 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
...
-#ifdef HOST_UNIX
-    return nullptr;
-#else
-    char buff[512 + sizeof(CORINFO_String)];
-
-    // make this bulletproof, so it works even if we are wrong.
-    if (ReadProcessMemory(GetCurrentProcess(), (void*)strHandle, buff, 4, nullptr) == 0)
+    const int maxStrSize = 64;
+    char      str[maxStrSize];
+    int       realLength = this->info.compCompHnd->objectToString((void*)handle, str, maxStrSize);
```
Comment (jkotas) [inline src/coreclr/jit/ee_il_dll.cpp:1622]: The managed impl returns "bytes written" in . Should it be bytes written or real length?

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 2]
```diff
@@ -1830,6 +1831,21 @@ private int getStringLiteral(CORINFO_MODULE_STRUCT_* module, uint metaTOK, char*
             return str.Length;
         }
 
+#pragma warning disable CA1822 // Mark members as static
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:1834]: This warning disable should not be needed

File: src/coreclr/jit/ee_il_dll.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/ee_il_dll.cpp:1622]: Ah, I guess it's a left-over from the initial impl that used to return real length (in utf16 chars) but then we switched to Utf8 for this API as it's only inteded to be used for debug/diagnostic purposes

Comment (jkotas) [reply src/coreclr/jit/ee_il_dll.cpp:1622]: > it's only intended to be used for debug/diagnostic purposes so it should be bytesWritten now

Add a comment to corinfo.h?

It would be nice to follow a consistent naming and behavior pattern for these diagnostic-only methods that return diagnostic strings. We have `objectToString` and `appendClassName` that @jakobbotsch is touching in a separate PR.

Comment (jakobbotsch) [reply src/coreclr/jit/ee_il_dll.cpp]: Personally, I very much dislike the "append" part of `appendClassName`. The caller can do this easily. If I had to pick I would probably pick:
1. The API always null terminates on its own
2. Lengths should be `size_t`
3. The API returns the actual length of a buffer necessary to hold the full result (including null terminator).
4. The API guarantees that `min(returnedSize, bufferSize)` characters were written to the buffer, including a null terminator.

I can move `appendClassName` to this model in my PR. I think we can name the APIs `printFrozenObject` and `printClassName`?

Any comments/suggestions @jkotas?

Comment (jakobbotsch) [reply src/coreclr/jit/ee_il_dll.cpp]: I can also move `getMethodName` to this model in my PR. Today it has different behavior in checked and release builds, and JIT is actually only using it for printing helper function names.

Comment (jakobbotsch) [reply src/coreclr/jit/ee_il_dll.cpp]: I think the SPMI recording could then be something like:
```cpp
Agnostic_AppendClassNameIn key;
key.Handle = classHandle;

size_t writtenSize = min(returnedSize, bufferSize);
int index = AppendClassName->GetIndex(key);
if (index != -1)
{
    Agnostic_AppendClassNameOut value = AppendClassName->GetItem(index);
    if (value.storedBufferLength >= bufferSize)
        return; // already storing more
}

Agnostic_AppendClassNameOut value;
value.returnedSize = returnedSize;
value.storedBufferSize = writtenSize;
value.bufferIndex = (DWORD)AppendClassName->AddBuffer(buffer, writtenSize);

if (index != -1)
    AppendClassName->Update(index, value);
else
    AppendClassName->Add(key, value);
```

that ensures we always store the largest call we saw.

Comment (jkotas) [reply src/coreclr/jit/ee_il_dll.cpp]: > Personally, I very much dislike the "append" part of appendClassName

I agree. The "append" part does not make sense.

> size_t printClassName(CORINFO_CLASS_HANDLE, char* buffer, size_t bufferSize)
> size_t printFrozenObject(void* handle, char* buffer, size_t bufferSize)

We should be consistent in whether we use "Object" or "FrozenObject" in the JIT-EE interface names. https://github.com/dotnet/runtime/pull/76112/files#diff-09491c1d7d4c7b20ad241d23d2b2ec182368b851cfb5f4ed4cdb5509dd6dd01fR291-R294 uses Object. I am not sure which one of the two options is better.

Comment (EgorBo) [reply src/coreclr/jit/ee_il_dll.cpp]: @jakobbotsch can you explain the spmi part, what's wrong with the current code?

Comment (jkotas) [reply src/coreclr/jit/ee_il_dll.cpp]: I like `print` as the prefix to use for methods that are meant to return textual representation for logging or diagnostic purposes.

Comment (jakobbotsch) [reply src/coreclr/jit/ee_il_dll.cpp]: I think it should work fine now that I look at it. Keying on the buffer size is somewhat odd, but hardly a problem given that we don't expect to see this JIT-EE function being used with many (or any) different buffer sizes.

The `appendClassName` version will need to be changed (it does not key on buffer size) so I'll try to factor them at that point.

Comment (jkotas) [reply src/coreclr/jit/ee_il_dll.cpp]: Also, we should be consistent in whether these methods zero-terminate the string when it does not fit or not. Either behavior is fine with me as long as it is consistent.

File: src/coreclr/jit/ee_il_dll.cpp [hunk 3]
```diff
@@ -1615,45 +1615,27 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
...
-        // strHandle is a frozen string
-        // We assume strHandle is never an "interior" pointer in a frozen string
-        // (jit is not expected to perform such foldings)
-        asString = (CORINFO_String*)strHandle;
-    }
-    else
+    const size_t maxStrSize = 64;
+    char         str[maxStrSize];
+    size_t       actualLen = this->info.compCompHnd->printObject((void*)handle, str, maxStrSize);
+    if (actualLen == -1)
```
Comment (jkotas) [inline src/coreclr/jit/ee_il_dll.cpp:1623]: `size_t` is unsigned. I expect that you will see some compilers complaining about it on some platforms.

I think it would be better if `print` APIs never fail and always give you back something.

Comment (EgorBo) [reply src/coreclr/jit/ee_il_dll.cpp]: Thanks for noticing, fixed. Now it will be just an empty buffer

File: src/coreclr/jit/ee_il_dll.cpp [hunk 4]
```diff
@@ -1615,45 +1615,27 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
...
-    {
-        // strHandle is a frozen string
-        // We assume strHandle is never an "interior" pointer in a frozen string
-        // (jit is not expected to perform such foldings)
-        asString = (CORINFO_String*)strHandle;
-    }
-    else
+    const size_t maxStrSize = 64;
+    char         str[maxStrSize];
+    size_t       actualLen = this->info.compCompHnd->printObject((void*)handle, str, maxStrSize);
```
Comment (jkotas) [inline src/coreclr/jit/ee_il_dll.cpp:1622]: > eePrintObjectDescription vs printObject

I would be nice if the internal JIT helper method and the JIT/EE interface method have same style name.

I have mild preference for `printObjectDescription`. It is more consistent with `printClassName` and `printMethodName`.

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 5]
```diff
@@ -1830,6 +1831,28 @@ private int getStringLiteral(CORINFO_MODULE_STRUCT_* module, uint metaTOK, char*
...
+            {
+                // Trim data and null-terminate it
+                bufferSpan[bufferSize32 - 1] = 0;
+            }
+            else
+            {
+                bufferSpan[written] = 0;
+            }
+
+            return (nuint)objStr.Length;
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:1853]: This does not match the documented contract. This is the UTF-16 length. The contract says that it should be the full UTF-8 length.

Comment (EgorBo) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs]: Oops, indeed, fixed

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 6]
```diff
@@ -1830,6 +1831,27 @@ private int getStringLiteral(CORINFO_MODULE_STRUCT_* module, uint metaTOK, char*
...
+            if (written >= bufferSize32)
+            {
+                // Trim data and null-terminate it
+                bufferSpan[bufferSize32 - 1] = 0;
+            }
+            else
+            {
+                bufferSpan[written] = 0;
+            }
+            return (nuint)Encoding.UTF8.GetByteCount(objStr);
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:1852]: Is it that expensive to do the simple easy thing with `byte[] utf8 = Encoding.UTF8.GetBytes(objStr)` from the beginning?

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs]: Changed, thanks

File: src/coreclr/inc/corinfo.h [hunk 7]
```diff
@@ -2265,6 +2265,27 @@ class ICorStaticInfo
...
+    //    purposes such as JitDisasm. The buffer is null-terminated (even if truncated).
+    //
+    // Arguments:
+    //    handle     - Direct object handle
+    //    buffer     - Pointer to buffer
+    //    bufferSize - Pointer to buffer length. Must not be nullptr
+    //
+    // Return Value:
+    //    Full length of the source UTF8 representation so can be bigger than bufferSize.
+    //    Includes null-terminator.
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:2281]:
```suggestion
    //    Full length of the source UTF8 representation so can be bigger than bufferSize.
    //    Includes null-terminator.
```
This is somewhat unusual convention for an API that takes buffer and returns `size_t` (e.g. sprintf_s). Is there a standard C API that has the same convention?

Comment (jkotas) [reply src/coreclr/inc/corinfo.h:2281]: `getStringLiteral` API is not pretty printing API. I do not see a problem with it. (BTW: We may want to change the `getStringLiteral` signature to take offset argument, so that we do not need to copy up to 1024 characters and immediately throw away most of it except the one character. Separate PR.)

I was commenting on the fact that it returns the length of the string plus the null terminator. I believe that the usual convention is to return the number of characters written into the buffer without the null terminator. `sprintf` is example of an API that follows this convention.

I do not have a preference between `size_t` and `int`. `size_t` is fine with me.

Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2281]: > so that we do not need to copy up to 1024 characters and immediately throw away most of it except the one character. 

That place definitely needs smaller buffer (or rather no buffer at all) no idea how it ended up with 1024, probably was used in JITDUMP initially.
It's also used on other paths where it consumes more characters e.g. StartsWith/Equals unrolling.

>  I believe that the usual convention is to return the number of characters written into the buffer without the null terminator

I agree, can change that in a moment, although, the advantage of the current API is that it easily tells you that you might want to try again with a larger buffer if you want to print the whole thing (so you can heap-allocate next chunk to the exact size) while with `bytesWritten` it's not clear -- was the string exactly 64 bytes? how many bytes should I try next attempt? 

But it all doesn't matter for this particular API since it's only supposed to be used in debug purposes and doesn't need to print the whole thing if it's too long.

Comment (jakobbotsch) [reply src/coreclr/inc/corinfo.h:2281]: Excluding null terminator is also fine with me, it seems better if that is more in line with the sprintf_s family of APIs. I still think we should return the number of bytes in the actual string (i.e., calling again with a buffer of sized `returnedSize + 1` will get the full string).

Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2281]: I assume in general there is no a standardized shape, e.g. APIs which work on streamed data can not know the final length in advance (a good example is that `Utf8.FromUtf16` that returns status of a operation e.g. `DestinationTooSmall`)

File: src/coreclr/inc/corinfo.h [hunk 8]
```diff
@@ -2265,6 +2265,27 @@ class ICorStaticInfo
...
+    //    object to a preallocated buffer. It's intended to use only for debug/diagnostic 
+    //    purposes such as JitDisasm. The buffer is null-terminated (even if truncated).
+    //
+    // Arguments:
+    //    handle     - Direct object handle
+    //    buffer     - Pointer to buffer
+    //    bufferSize - Pointer to buffer length. Must not be nullptr
+    //
+    // Return Value:
+    //    Full length of the source UTF8 representation so can be bigger than bufferSize.
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:2280]: This comment needs updating too

Comment (EgorBo) [reply src/coreclr/inc/corinfo.h]: Oops, done!

File: src/coreclr/inc/corinfo.h [hunk 9]
```diff
@@ -2277,8 +2277,7 @@ class ICorStaticInfo
     //    bufferSize - Pointer to buffer length. Must not be nullptr
     //
     // Return Value:
-    //    Full length of the source UTF8 representation so can be bigger than bufferSize.
-    //    Includes null-terminator.
+    //    Bytes written to the given buffer, the range is [0..bufferSize]
```
Comment (jakobbotsch) [inline src/coreclr/inc/corinfo.h:2280]: I'd like the JIT to be able to print the exact class name even for large lengths, so this means `printClassName` is going to be inconsistent with this API design.

I guess we can use a doubling strategy, but it seems unnecessary to me.

Comment (jkotas) [reply src/coreclr/inc/corinfo.h:2280]: I would add an extra `size* pRequiredBufferSize = nullptr` argument for this. It allows you to get the full buffer size only if you actually want it.

File: src/coreclr/inc/corinfo.h [hunk 10]
```diff
@@ -2265,6 +2265,26 @@ class ICorStaticInfo
...
+
+    //------------------------------------------------------------------------------
+    // printObjectDescription: Prints a (possibly truncated) textual UTF8 representation of the given
+    //    object to a preallocated buffer. It's intended to be use only for debug/diagnostic 
+    //    purposes such as JitDisasm. The buffer is null-terminated (even if truncated).
+    //
+    // Arguments:
+    //    handle     - Direct object handle
+    //    buffer     - Pointer to buffer
+    //    bufferSize - Pointer to buffer length. Must not be nullptr
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:2277]:
```suggestion
    //    bufferSize - Buffer size
```

File: src/coreclr/inc/corinfo.h [hunk 11]
```diff
@@ -2265,6 +2265,26 @@ class ICorStaticInfo
             int                         bufferSize  /* IN  */
             ) = 0;
 
+
+    //------------------------------------------------------------------------------
+    // printObjectDescription: Prints a (possibly truncated) textual UTF8 representation of the given
+    //    object to a preallocated buffer. It's intended to be use only for debug/diagnostic
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:2271]:
```suggestion
    //    object to a preallocated buffer. It's intended to be used only for debug/diagnostic 
```

File: src/coreclr/inc/corinfo.h (hunk 9 above)
Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2280]: Do you want me to add an additional argument here just for consistency or @jakobbotsch you plan to rely on it? Extending JIT-EE is quite painful 😞

Comment (jkotas) [reply src/coreclr/inc/corinfo.h:2280]: I would add it so that we have full flexibility and do not have to deal with this ever again. @jakobbotsch ?

Comment (jakobbotsch) [reply src/coreclr/inc/corinfo.h:2280]: Sounds fine to me.

Nit: the returned value is going to be in range `[0..bufferSize)` if it does not include null terminator.

Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2280]: It does include it so should be `bufferSize]`

Comment (jakobbotsch) [reply src/coreclr/inc/corinfo.h:2280]: I was under the impression we had converged on not including the null terminator to be in line with `sprintf_s` family of functions?

Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2280]: I am not sure I understand the final decision either - a bit lost in the comments, if we don't want to null-terminate it then we basically are back to the shape that was before we start changing it back and forth 😄 except the `size* pRequiredBufferSize` parameter

Comment (jakobbotsch) [reply src/coreclr/inc/corinfo.h:2280]: The API is null terminating always, just the return value does not include the null terminator. Then it is consistent with C style string functions.

Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2280]: Or do we want to still null-terminate it there but not include the `\0` into the return value? Ok

Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2280]: Ah, I see

File: src/coreclr/vm/jitinterface.cpp [hunk 12]
```diff
@@ -715,6 +715,71 @@ int CEEInfo::getStringLiteral (
...
+        buffer[bytesWritten - 1] = 0;
+    }
+    else
+    {
+        buffer[bytesWritten] = 0;
+    }
+
+    if (pRequiredBufferSize != nullptr)
+    {
+        *pRequiredBufferSize = stackStr.GetCount();
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:775]:
```suggestion
        *pRequiredBufferSize = stackStr.GetCount() + 1;
```
So that the output fits into the buffer.

Comment (EgorBo) [reply src/coreclr/vm/jitinterface.cpp:775]: Ah, right, I had "real length" here in mind

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 13]
```diff
@@ -1830,6 +1831,30 @@ private int getStringLiteral(CORINFO_MODULE_STRUCT_* module, uint metaTOK, char*
...
+                // Trim data and null-terminate it
+                bufferSpan[written - 1] = 0;
+            }
+            else
+            {
+                bufferSpan[written] = 0;
+            }
+            if (pRequiredBufferSize != null)
+            {
+                *pRequiredBufferSize = (nuint)Encoding.UTF8.GetByteCount(objStr);
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs]:
```suggestion
                *pRequiredBufferSize = (nuint)Encoding.UTF8.GetByteCount(objStr) + 1;
```

Comment (EgorBo) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs]: Thanks, already applied by hands

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 14]
```diff
@@ -1830,6 +1831,30 @@ private int getStringLiteral(CORINFO_MODULE_STRUCT_* module, uint metaTOK, char*
...
+            Utf8.FromUtf16(objStr, bufferSpan, out _, out int written);
+            if (written == bufferSize32)
+            {
+                // Trim data and null-terminate it
+                bufferSpan[written - 1] = 0;
+            }
+            else
+            {
+                bufferSpan[written] = 0;
+            }
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:1850]:
```suggestion
            if (bufferSize > 0)
            {
                Utf8.FromUtf16(objStr, new Span<byte>(buffer, checked((int)(bufferSize - 1))), out _, out int written);
                // Always null-terminate
                buffer[written] = 0;
        }
```

Comment (EgorBo) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs]: Thanks, nice clean up!

**Review by jkotas — APPROVED** [review]
:shipit: 

(I have not reviewed the superpmi files.)

File: src/coreclr/vm/jitinterface.cpp [hunk 15]
```diff
@@ -715,6 +715,67 @@ int CEEInfo::getStringLiteral (
...
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    size_t bytesWritten = 0;
+
+    // NOTE: this function is used for pinned/frozen handles
+    // it doesn't need to null-terminate the string
+
+    _ASSERT(handle != nullptr && buffer != nullptr && bufferSize > 0);
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:735]:
```suggestion
    _ASSERT(handle != nullptr);
```

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 16]
```diff
@@ -1830,6 +1831,26 @@ private int getStringLiteral(CORINFO_MODULE_STRUCT_* module, uint metaTOK, char*
             return str.Length;
         }
 
+        private nuint printObjectDescription(void* handle, byte* buffer, nuint bufferSize, nuint* pRequiredBufferSize)
+        {
+            Debug.Assert(bufferSize > 0 && handle != null && buffer != null);
+
+            int bufferSize32 = checked((int)bufferSize);
+            ReadOnlySpan<char> objStr = HandleToObject(handle).ToString();
+            var bufferSpan = new Span<byte>(buffer, bufferSize32);
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:1840]:
```suggestion
```

File: src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp [hunk 17]
```diff
@@ -4875,6 +4898,81 @@ int MethodContext::repGetStringLiteral(CORINFO_MODULE_HANDLE module, unsigned me
...
+        return 0;
+    }
+    else
+    {
+        Agnostic_PrintObjectDescriptionResult value = PrintObjectDescription->Get(key);
+        DEBUG_REP(dmpPrintObjectDescription(key, value));
+        if (pRequiredBufferSize != nullptr)
+        {
+            *pRequiredBufferSize = (size_t)value.requiredBufferSize;
+        }
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:4953]: @jakobbotsch I am not sure I understand why it needs such complications on jit side I don't see us doing that for all SPMI APIs we invoke

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:4953]: especially for just a poor debug-only api

File: src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp [hunk 18]
```diff
@@ -4875,6 +4898,81 @@ int MethodContext::repGetStringLiteral(CORINFO_MODULE_HANDLE module, unsigned me
...
+        PrintObjectDescription = new LightWeightMap<DLDL, Agnostic_PrintObjectDescriptionResult>();
+
+    DLDL key;
+    key.A = CastHandle(handle);
+    key.B = (DWORDLONG)bufferSize;
+
+    DWORD strBuf = (DWORD)-1;
+    if (buffer != nullptr && bytesWritten > 0)
+    {
+        size_t bufferRealSize = min(bytesWritten, bufferSize);
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:4913]: Can you explain this change - I don't understand it

File: src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:4953]: Also, `pRequiredBufferSize` can be null, in fact it is null always currently

[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:4953]: > @jakobbotsch I am not sure I understand why it needs such complications on jit side I don't see us doing that for all SPMI APIs we invoke

For APIs that we do not care if they fail we are using SPMI error trap on the JIT side normally, e.g. see `eeGetMethodName`, `eeGetFieldName` and `eeTryGetClassSize`. So normally we will make the SPMI implementation fail on missing data and let JIT decide on whether it cares or not. Since I am going to be factoring this code into some common helper function I will probably end up switching it to fail on missing data (in particular we want this behavior for `printMethodName` so we can detect it on the JIT side).

> Also, pRequiredBufferSize can be null, in fact it is always null currently

What I mean is: if `pRequiredBufferSize` is non-null, but the data was not captured during collection. It does not happen today but like the collection/replay to be robust to situations like that.
I understand that this API is only used for diagnostic cases and in those scenarios we don't really care what gets returned here on missing data. It is not going to be the case for `printMethodName` that we need for `JitDisasm`/`JitDump`.

File: src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:4913]: It seems odd to me that in the right case we are storing the null terminator while in the left case we do not store the null terminator.
In reality `bytesWritten` is always going to be less than `bufferSize` though, so it doesn't matter.

File: src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp [hunk 19]
```diff
@@ -4875,6 +4898,81 @@ int MethodContext::repGetStringLiteral(CORINFO_MODULE_HANDLE module, unsigned me
...
+            if (bytesWritten >= bufferSize)
+            {
+                // Trim data and null-terminate it
+                buffer[bytesWritten - 1] = 0;
+            }
+            else
+            {
+                buffer[bytesWritten] = 0;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:4971]: I am not even sure why we need all of that in SPMI that is supposed to just replay what a vm put into that buffer

[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp]: Doing it that way would be fine with me too (but then you need to ensure the buffer is recorded with the null terminator, and the `bytesWritten > 0` check above needs some changes)

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp]: Yep, I copied Jan's trick here
